File: app/schemas/sec_filings.py
"""SEC filing 列表与分类 Pydantic schemas。"""

from typing import List

from pydantic import Field

from app.schemas.base import BaseResponse


class EightKItem(BaseResponse):
    code: str = Field(description="8-K item 代码，如 2.02")
    label: str = Field("", description="item 中文释义，未知代码为空")


class FilingRecord(BaseResponse):
    form: str = Field(description="SEC 表格类型，如 10-K / 8-K / 4")
    category: str = Field(description="所属分类 key")
    filing_date: str = Field(description="提交日 YYYY-MM-DD")
    report_date: str = Field("", description="报告期日 YYYY-MM-DD（可能与提交日不同，也可能为空）")
    accession_number: str = Field(description="SEC accession number")
    primary_doc_description: str = Field("", description="主文档描述")
    items: List[EightKItem] = Field(default_factory=list, description="8-K item 明细，其他表格为空")
    index_url: str = Field("", description="SEC filing 目录页链接")
    doc_url: str = Field("", description="主文档直链")


class FilingCategory(BaseResponse):
    key: str = Field(description="分类 key")
    label: str = Field(description="分类中文标签")
    label_en: str = Field(description="分类英文标签")
    count: int = Field(description="该分类下 filing 数量")
    filings: List[FilingRecord] = Field(default_factory=list, description="该分类的 filing 列表（新→旧）")


class CompanyInfo(BaseResponse):
    cik: str = Field(description="10 位补零 CIK")
    name: str = Field("", description="公司名称")
    tickers: List[str] = Field(default_factory=list, description="股票代码列表")
    exchanges: List[str] = Field(default_factory=list, description="交易所列表")
    sic_description: str = Field("", description="行业 SIC 描述")


class CompanyFilingsResponse(BaseResponse):
    company: CompanyInfo
    total: int = Field(description="全部 filing 总数")
    categories: List[FilingCategory] = Field(default_factory=list, description="按分类分组的 filing")

File: frontend/lib/api/sec_filings.ts
// frontend/lib/api/sec_filings.ts
import apiClient from './client'

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

export interface EightKItem {
  code: string
  label: string
}

export interface FilingRecord {
  form: string
  form_name: string
  form_desc: string
  category: string
  filing_date: string
  report_date: string
  accession_number: string
  primary_doc_description: string
  items: EightKItem[]
  index_url: string
  doc_url: string
}

export interface FilingCategory {
  key: string
  label: string
  label_en: string
  count: number
  filings: FilingRecord[]
}

export interface CompanyInfo {
  cik: string
  name: string
  tickers: string[]
  exchanges: string[]
  sic_description: string
}

export interface CompanyFilingsResponse {
  request_id: string
  company: CompanyInfo
  total: number
  categories: FilingCategory[]
}

export async function fetchCompanyFilings(
  query: string
): Promise<CompanyFilingsResponse> {
  const { data } = await apiClient.get<CompanyFilingsResponse>('/api/v1/sec/filings', {
    params: { query },
    timeout: 60000, // SEC 溢出文件抓取可能较慢
  })
  return data
}

export interface FilingDocument {
  seq: string
  type: string
  label: string
  description: string
  filename: string
  url: string
  highlight: boolean
}

export interface FilingDocumentsResponse {
  request_id: string
  accession: string
  index_url: string
  documents: FilingDocument[]
}

export async function fetchFilingDocuments(
  cik: string,
  accession: string
): Promise<FilingDocumentsResponse> {
  const { data } = await apiClient.get<FilingDocumentsResponse>('/api/v1/sec/filing-documents', {
    params: { cik, accession },
    timeout: 30000,
  })
  return data
}

export interface ProductItem {
  name: string
  explanation: string
  market_share: string
}

export interface CompanyProfile {
  one_liner: string
  industry: string
  supply_chain_position: string
  main_products: ProductItem[]
  main_customers: string[]
  moat_rating: '宽' | '中' | '窄' | '无'
  moat_reason: string
  differentiation: string
  competitors: string[]
}

export interface CompanyProfileResponse {
  request_id: string
  cik: string
  name: string
  ticker: string
  sic_description: string
  profile: CompanyProfile
}

export type CompanyProfileStreamEvent =
  | {
      event: 'start' | 'meta' | 'generating'
      message: string
      company?: Partial<CompanyProfileResponse>
      done?: false
    }
  | {
      event: 'resolved' | 'cache_hit'
      message: string
      company: Partial<CompanyProfileResponse>
      done?: false
    }
  | {
      event: 'done'
      data: CompanyProfileResponse
      done: true
    }
  | {
      event: 'error'
      message: string
      done: true
    }

export async function fetchCompanyProfile(
  query: string
): Promise<CompanyProfileResponse> {
  const { data } = await apiClient.get<CompanyProfileResponse>('/api/v1/sec/company-profile', {
    params: { query },
    timeout: 60000, // 大模型生成可能较慢
  })
  return data
}

export async function* streamCompanyProfile(
  query: string,
  signal?: AbortSignal,
): AsyncGenerator<CompanyProfileStreamEvent> {
  const params = new URLSearchParams({ query })
  const headers: HeadersInit = { Accept: 'text/event-stream' }
  if (typeof window !== 'undefined') {
    const token = localStorage.getItem('access_token')
    if (token) headers.Authorization = `Bearer ${token}`
  }

  const response = await fetch(`${API_BASE_URL}/api/v1/sec/company-profile/stream?${params.toString()}`, {
    headers,
    signal,
  })

  if (!response.ok) {
    throw new Error(`公司概览生成失败 (${response.status})`)
  }
  if (!response.body) {
    throw new Error('浏览器不支持流式响应')
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    const chunks = buffer.split('\n\n')
    buffer = chunks.pop() ?? ''

    for (const chunk of chunks) {
      const dataLine = chunk
        .split('\n')
        .find((line) => line.startsWith('data: '))
      if (!dataLine) continue
      yield JSON.parse(dataLine.slice(6)) as CompanyProfileStreamEvent
    }
  }

  buffer += decoder.decode()
  if (buffer.trim()) {
    const dataLine = buffer
      .split('\n')
      .find((line) => line.startsWith('data: '))
    if (dataLine) {
      yield JSON.parse(dataLine.slice(6)) as CompanyProfileStreamEvent
    }
  }
}
